### PR TITLE
split `restore_training_state` into logical parts [1 / 2]

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

In #7900 `CheckpointConnector.restore_training_state` gets split into multiple pieces:

```
restore_callbacks()
restore_progress()
restore_optimizers()
restore_lr_schedulers()
```

This PR introduces the new functions, but they are unused. 
#7900 will then refactor the `CheckpointConnector.restore_training_state` method to use them. 

This is mainly to reduce a hard to read diff for reviews!

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lihtning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 1]
```diff
@@ -207,6 +207,93 @@ def restore_training_state(self, checkpoint, load_optimizer_states: bool = True)
         for scheduler, lrs_state in zip(self.trainer.lr_schedulers, lr_schedulers):
             scheduler['scheduler'].load_state_dict(lrs_state)
 
+    def restore_callbacks(self) -> None:
+        """ Restores all callbacks from the pre-loaded checkpoint. """
+        if not self._loaded_checkpoint:
+            return
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:213]: is there a risk of these restoration functions being called outside this context? should the start and end restore from checkpoint be replaced with a dedicated context manager for resuming?

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 2]
```diff
@@ -207,6 +207,93 @@ def restore_training_state(self, checkpoint, load_optimizer_states: bool = True)
...
+
+        if any([key in self._loaded_checkpoint for key in DEPRECATED_CHECKPOINT_KEYS]):
+            raise ValueError(
+                "The checkpoint you're attempting to load follows an"
+                " outdated schema. You can upgrade to the current schema by running"
+                " `python -m pytorch_lightning.utilities.upgrade_checkpoint --file model.ckpt`"
+                " where `model.ckpt` is your checkpoint file."
+            )
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:221]: should this validation be done in `resume_start` ?

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:213]: Hey, good question. 
If you look at the "end result" #7652 (open to discussion) you will see here in the Trainer file resume_start() and resume_end are actually in very different places(), so I can't make it into a context manager.

Yes, I think it's best to document the order. The order may be important. In the future, we will want to enable configuration of what is restored, so some of these functions get called on demand and some won't be called.

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 2 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:221]: good question. maybe we could. 
one though though, in the future we will have a way to configure what to load, so if these functions get called individually we may want to have the validation together with the particular objects that are being restored.

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:213]: actually, maybe a context manager could still work. I will investigate it in #7652

Comment (carmocca) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:213]: I think Ananth's suggestion is good.

Also, could any other class want to call the `start` and `end` methods?

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 3]
```diff
@@ -207,6 +207,93 @@ def restore_training_state(self, checkpoint, load_optimizer_states: bool = True)
...
+
+        self.trainer.train_loop.global_step = self._loaded_checkpoint['global_step']
+        self.trainer.train_loop.current_epoch = self._loaded_checkpoint['epoch']
+
+        # crash if max_epochs is lower then the current epoch from the checkpoint
+        if self.trainer.max_epochs is not None and self.trainer.current_epoch > self.trainer.max_epochs:
+            m = f"""
+            you restored a checkpoint with current_epoch={self.trainer.current_epoch}
+            but the Trainer(max_epochs={self.trainer.max_epochs})
+            """
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:240]: I imagine you just copy-pasted this but it'd be nice if you could convert it to a "normal" string

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 4]
```diff
@@ -207,6 +207,93 @@ def restore_training_state(self, checkpoint, load_optimizer_states: bool = True)
...
+        """ Restores the optimizers and learning rate scheduler states from the pre-loaded checkpoint. """
+        if not self._load_optimizer_states or not self._loaded_checkpoint:
+            return
+
+        # validation
+        if "optimizer_states" not in self._loaded_checkpoint or "lr_schedulers" not in self._loaded_checkpoint:
+            raise KeyError(
+                "Trying to restore training state but checkpoint contains only the model."
+                " This is probably due to `ModelCheckpoint.save_weights_only` being set to `True`."
+            )
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:265]: Do this chain of PRs plan to tackle the issue of restoring part of the checkpoint?

https://github.com/PyTorchLightning/pytorch-lightning/issues/5339

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 4 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:265]: not the aim directly but will definitely help, and we can directly continue with it after these PRs. There will be nothing standing in the way as far as I can tell :)

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 3 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py]: good catch, ok I sneaked it in.

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:213]: No, I think we would only want to call it for unit testing, or the context manager if that works. 
So I know what you are saying, yes I will put the underscores everywhere

Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:213]: okay, ctx manager could kinda work but I see an issue. can we move the conversation to #7652 so I can directly point to the code in trainer.py?

**Review by tchaton — APPROVED** [review]
Love it ! Really nice cleanup.

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 5]
```diff
@@ -207,6 +207,92 @@ def restore_training_state(self, checkpoint, load_optimizer_states: bool = True)
         for scheduler, lrs_state in zip(self.trainer.lr_schedulers, lr_schedulers):
             scheduler['scheduler'].load_state_dict(lrs_state)
 
+    def restore_callbacks(self) -> None:
+        """ Restores all callbacks from the pre-loaded checkpoint. """
+        if not self._loaded_checkpoint:
+            return
+
+        if any([key in self._loaded_checkpoint for key in DEPRECATED_CHECKPOINT_KEYS]):
```
Comment (Borda) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:215]:
```suggestion
        if any(key in self._loaded_checkpoint for key in DEPRECATED_CHECKPOINT_KEYS):
```

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 6]
```diff
@@ -207,6 +207,92 @@ def restore_training_state(self, checkpoint, load_optimizer_states: bool = True)
...
+            return
+
+        self.trainer.train_loop.global_step = self._loaded_checkpoint['global_step']
+        self.trainer.train_loop.current_epoch = self._loaded_checkpoint['epoch']
+
+        # crash if max_epochs is lower then the current epoch from the checkpoint
+        if self.trainer.max_epochs is not None and self.trainer.current_epoch > self.trainer.max_epochs:
+            raise MisconfigurationException(
+                f"You restored a checkpoint with current_epoch={self.trainer.current_epoch},"
+                f" but the Trainer(max_epochs={self.trainer.max_epochs})"
```
Comment (ethanwharris) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:239]: Looks like this error message is missing a bit

**Review by ethanwharris — APPROVED** [review]
LGTM, small comment

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 6 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:239]: yeah, it was like this already ^^

I could change the text to:

... but you have set Trainer(max_epochs=...)"

Comment (ethanwharris) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py]: Haha, yeah, that would do :smiley:
